### ISSUE TITLE
Rule proposal: "no-misused-collection" warn against using `in` on `Set` and `Map`

### ISSUE THREAD

**Issue by bradzacher** [typescript-eslint/typescript-eslint#5677]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

In languages like [python](https://www.w3schools.com/python/gloss_python_check_if_set_item_exists.asp) use the `in` operator to check for key existence in sets and maps. Polyglot engineers may make the mistake to use `in` in JS.

With [this change to TS to be released in 4.9](https://github.com/microsoft/TypeScript/pull/50666), the `in` operator will now refine any object type by intersecting it with `Record<key, unknown>`. This means you could get some funky behaviour that previously was not allowed.
For example:

```ts
declare const map: Map<string, unknown>;

declare function acceptsAnything(arg: unknown): void;

if ('woopsie' in map) {
  acceptsAnything(map['woopsie']);
}
```
[Currently in TS 4.8 this errors with `Element implicitly has an 'any' type because type 'Map<string, unknown>' has no index signature. Did you mean to call 'map.get'? (7052)`](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEYD2A7AzgF3gWygBwC54BZfAHkxgEsUBzAGngFcUBrFJAdxQD4BuAFADQkWAgBmLMBiqp4UMGBB4MaAIIoAnhgAWNWgApYtIi3ZcUASiIA3JFWCCBVcfAMByTkiR40VEO7wNDj4lvAA3gLw8orKqhraenQGuHgA2p7evv7uALqWggC+AkA)
[After the above PR, this no longer errors at all](https://www.staging-typescript.org/play?ts=4.9.0-pr-50666-16#code/CYUwxgNghgTiAEYD2A7AzgF3gWygBwC54BZfAHkxgEsUBzAGngFcUBrFJAdxQD4BuAFADQkWAgBmLMBiqp4UMGBB4MaAIIoAnhgAWNWgApYtIi3ZcUASiIA3JFWCCBVcfAMByTkiR40VEO7wNDj4lvAA3gLw8orKqhraenQGuHgA2p7evv7uALqWggC+AkA) because the type of `map` is refined to `Map<string, unknown> & Record<'woopsie', unknown>`, so `typeof map['woopsie'] === unknown`.

### Fail Cases

```typescript
declare const set: Set<string>;
if ('woopsie' in set) {
}


declare const map: Map<string, unknown>;
if ('woopsie' in map) {
}

map['woopsie'] = 1;
map['woopsie'].length;
```

### Pass Cases

```typescript
declare const set: Set<string>;
set.has('woopsie');
set.add('woopsie');
set.delete('woopsie');


declare const map: Map<string, unknown>;
map.has('woopsie');
map.get('woopsie');
map.set('woopsie', 1);
map.delete('woopsie');
```

**Comment by Josh-Cena** [comment]
What about "no-misused-in"?

**Comment by bradzacher** [comment]
If there are more types that we think it's invalid to use `in` on, perhaps this could be expanded to `no-misused-in`.

Perhaps it's worth just making this error on `in` usage with any type other than (explicitly) `object`, `{}` or `Record`.

**Comment by Josh-Cena** [comment]
`in` has been notoriously unsound to begin with, because TS doesn't forbid excess properties. People should not be using it to narrow unions either.

**Comment by bradzacher** [comment]
@Josh-Cena could you clarify with some example code how it would be unsafe to use with a union?

**Comment by Josh-Cena** [comment]
[Playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAcg9sAggFQIYHN0QCYFUB2AlnPlALwBQUUAPlAN5SoBcU+ArgLYBGEATgG4o3VgGdgfQvnRQAvlVoMoAY1Yce-ORQoAzdvmXBipTqgDWERPgDy3AFYRDACgCUreEjSYcBYwwUA9AFQAELgqKKiUBAAHsoQkVBgfHCQfKAqABaOZgrKJOJQcPbkSqpQAIwANMKsAExyAgp8EMDsfKTFdk3yFPn4hV2lphZWtg7OLk2EOlBOAETc81BSRfYu-tT9onAANhAAdLtw6E5dB6gHwHAAojFgJBD4Rqi7rlMU8kA)

**Comment by bradzacher** [comment]
Yeah I guess structural typing is a bitch there and it allows some dodgy code!
Worth noting that if you directly return the object instead of indirectly via a variable, then TS applies an excess property check.

**Comment by Josh-Cena** [comment]
Correct—but in an actual program it's rarely that trivial, and the `makeAnObject` function may come from a library and you never know what's inside it...

**Comment by HolgerJeromin** [comment]
When moving from js objects to `Map` we got many problems with incomplete migration which worked but was not intended.
Could this be handled with this item or should I open a new one?

```ts
this.__object = {};
// ...
this.__object['objectPropertyname'];
```
```ts
this.__map = new Map();
// ...
this.__map['objectPropertyname']; // ops
```

**Comment by bradzacher** [comment]
@HolgerJeromin - originally I was thinking that the best plan for this rule was ensuring correct usage of ES6 collections - but I think there's more value in specifically just handling the `in` operator given the upcoming change in TS 4.9.

Given that TS has a compiler option which would handle the index access usecase in collections - I don't think it makes sense for us to build a rule in this plugin for it.
Though it would be pretty straight-forward to build one for your codebase using our tooling!

**Comment by mahdi-farnia** [comment]
I just wanna get a background on **what is exactly is a misuse of `in` operator.**

I wanna work on this, It seems complex for me as a new one in ESLint & ast... (But I'll try)

I know one of them which is type `Array<any>`, what should be added?

**Comment by bradzacher** [comment]
@mahdi-farnia does the issue's  description contain enough information to answer your question?

**Comment by mahdi-farnia** [comment]
@bradzacher
At time that i wrote that comment, I didn't read your comment about `__index` and ts-ast viewer. ( about no-in-array ).

Know i do.

So **Yes**!

**Comment by karlhorky** [comment]
I was going to submit a new rule proposal just now for `no-unsafe-in`, following the existing rules:

- [`@typescript-eslint/no-unsafe-argument`](https://typescript-eslint.io/rules/no-unsafe-argument)
- [`@typescript-eslint/no-unsafe-assignment`](https://typescript-eslint.io/rules/no-unsafe-assignment)
- [`@typescript-eslint/no-unsafe-call`](https://typescript-eslint.io/rules/no-unsafe-call)
- [`@typescript-eslint/no-unsafe-member-access`](https://typescript-eslint.io/rules/no-unsafe-member-access)
- [`@typescript-eslint/no-unsafe-return`](https://typescript-eslint.io/rules/no-unsafe-return)

Something that would ban usage of the `in` operator with `any`-typed RHS:

```ts
if ('x' in (undefined as any)) {
```

But maybe [comment #1251768783](https://github.com/typescript-eslint/typescript-eslint/issues/5677#issuecomment-1251768783) by @bradzacher alludes to `no-misused-in` already becoming this?

> Perhaps it's worth just making this error on `in` usage with any type other than (explicitly) `object`, `{}`, or a type with an index signature (eg `Record`).

**Comment by bradzacher** [comment]
I think that would be the best thing yeah - one rule for safe `in` usage.

**Comment by JoshuaKGoldberg** [comment]
Hmm, my instinct is that it'd be better to have two rules:

* `no-misused-in`: for the more nuanced case of `in` in objects
* `no-unsafe-in`: for banning `in` on `any`s as a very universal point of `any` safety

I think `no-misused-in` is a little too different from `no-unsafe-in` to implement in the same rule. Only `no-unsafe-in` actually deals with `any`. So it'd be weird for us to lump them in together IMO.

Thoughts @typescript-eslint/triage-team? I'm ready to implement this either way.
